src/gwt/src/org/rstudio/studio/client/workbench/prefs/views/GeneralPreferencesPane.java

```diff
@@ -518,6 +546,25 @@ private boolean getRestoreProjectRVersion()
          return false;
    }
    
+   private void restart()
+   {
+      quit_.prepareForQuit(
+            "Restarting RStudio",
+            new QuitContext()
+            {
+               @Override
+               public void onReadyToQuit(boolean saveChanges)
+               {
+                  String project = session_.getSessionInfo().getActiveProjectFile();
+                  if (project == null)
+                     project = Projects.NONE;
+                  
+                  events_.fireEvent(new OpenProjectNewWindowEvent(project, null));
+                  quit_.performQuit(null, saveChanges);
```

Comment:
If I'm reading this correctly, the new instance of the project will be opened *before* the current one is closed -- so might be too late to commit any changes in `performQuit`; the new copy of RStudio may have already consumed the state. 

In addition of course you're going to lose all your currently open tabs and other per-session state, which wouldn't happen with a manual restart. 

I'm not sure this is worth investing in since these settings are unlikely to be touched on a day-to-day basis, but if we want to make it work I suspect we're going to need to do something on the Qt side.

Reply:
I think this commit should address those concerns: https://github.com/rstudio/rstudio/pull/3545/commits/5af348e6e679af60e8b430dd43dca3e5c0cb2163

It ensures the new RStudio session is launched only after quit from the current session has been acknowledged and completed. Do you think that will suffice?

Reply:
It might! Do you get to keep your source database in this sequence?

Reply:
Indeed you do :-)